modules/mobile-agents/android/client/src/org/wso2/mdm/agent/services/Operation.java

```diff
@@ -0,0 +1,899 @@
+	}
+
+	/**
+	 * Executes device management operations on the device.
+	 * @param operationCode - Device management operation code.
+	 * @param operationData - Operation data if required(ex:WIFI configuration).
+	 * @param requestMode   - Request mode (GCM/Local).
+	 * @return Operation/Policy status list.
+	 */
+	public JSONArray doTask(String operationCode, String operationData) {
+
+		switch (operationCode) {
+			case Constants.OPERATION_DEVICE_INFO:
+				getDeviceInfo(operationCode);
+				break;
+			case Constants.OPERATION_DEVICE_LOCATION:
+				getLocationInfo(operationCode);
+				break;
+			case Constants.OPERATION_GET_APPLICATION_LIST:
+				getApplicationList(operationCode);
+				break;
+			case Constants.OPERATION_LOCK_DEVICE:
+				lockDevice(operationCode);
+				break;
+			case Constants.OPERATION_WIPE_DATA:
+				wipeDevice(operationCode, operationData);
+				break;
+			case Constants.OPERATION_CLEAR_PASSWORD:
+				clearPassword(operationCode);
+				break;
+			case Constants.OPERATION_NOTIFICATION:
+				displayNotification(operationCode);
+				break;
+			case Constants.OPERATION_WIFI:
+				configureWifi(operationCode, operationData);
+				break;
+			case Constants.OPERATION_DISABLE_CAMERA:
+				disableCamera(operationCode, operationData);
+				break;
+			case Constants.OPERATION_INSTALL_APPLICATION:
+				installAppBundle(operationCode, operationData);
+				break;
+			case Constants.OPERATION_INSTALL_APPLICATION_BUNDLE:
+				installAppBundle(operationCode, operationData);
+				break;
+			case Constants.OPERATION_UNINSTALL_APPLICATION:
+				uninstallApplication(operationCode, operationData);
+				break;
+			case Constants.OPERATION_ENCRYPT_STORAGE:
+				encryptStorage(operationCode, operationData);
+				break;
+			case Constants.OPERATION_RING:
+				ringDevice(operationCode);
+				break;
+			case Constants.OPERATION_MUTE:
+				muteDevice(operationCode);
+				break;
+			case Constants.OPERATION_WEBCLIP:
+				createWebClip(operationCode, operationData);
+				break;
+			case Constants.OPERATION_PASSWORD_POLICY:
+				setPasswordPolicy(operationCode, operationData);
+				break;
+			case Constants.OPERATION_INSTALL_GOOGLE_APP:
+				installGooglePlayApp(operationCode, operationData);
+				break;
+			case Constants.OPERATION_CHANGE_LOCK_CODE:
+				changeLockCode(operationCode, operationData);
+				break;
+			case Constants.OPERATION_POLICY_BUNDLE:
+				setPolicyBundle(operationCode);
+				break;
+			case Constants.OPERATION_ENTERPRISE_WIPE_DATA:
+				enterpriseWipe(operationCode);
+				break;
+			case Constants.OPERATION_BLACKLIST_APPS:
+				blacklistApps(operationCode, operationData);
+				break;
+			default:
+				Log.e(TAG, "Invalid operation code receieved");
+				break;
+		}
+
+		return resultBuilder.getResultPayload();
+	}
+
+	/**
+	 * Retrieve device information.
+	 * @param code - Operation code.
+	 */
+	public void getDeviceInfo(String code) {
+		DeviceInfo deviceInfo = new DeviceInfo(context.getApplicationContext());
```

Comment:
Can we not initialize and store DeviceInfo, DeviceState and GPSTracker within, probably, the constructor of the operation class and then re-use, rather than initializing them over and over again within the critical path of getDeviceInfo operation?